cai_causal_graph/identify_utils.py

```diff
@@ -43,43 +46,76 @@ def identify_confounders(graph: CausalGraph, node_1: str, node_2: str) -> List[s
+
+        # create a copy of the provided graph and prune the children of node 1 and node 2
+        final_graph = clean_graph.copy()
+        for child in final_graph.get_children(n1):
+            final_graph.remove_edge(source=n1, destination=child)
+        for child in final_graph.get_children(n2):
+            final_graph.remove_edge(source=n2, destination=child)
+
+        # search the parents of node 1 and check whether any of them is an ancestor to node 2
+        confounders = set()
+        for parent in final_graph.get_parents(n1):
+            # add the parent to the confounding set if a directed path exists
+            if parent in final_graph.get_ancestors(n2):
+                confounders.add(parent)
+            # otherwise, recursively call this function to identify confounders of the parent
+            else:
+                confounders = confounders.union(
+                    set(_identify_confounders_no_checks_no_descendant_pruning(final_graph, parent, n2))
+                )
+
+        # do the reverse of the above by searching through the parents of node 2
+        confounders_reverse = set()
+        for parent in final_graph.get_parents(n2):
+            # add the parent to the confounding set if a directed path exists
+            if parent in final_graph.get_ancestors(n1):
+                confounders_reverse.add(parent)
+            # otherwise, recursively call this function to identify confounders of the parent
+            else:
+                confounders_reverse = confounders_reverse.union(
+                    set(_identify_confounders_no_checks_no_descendant_pruning(final_graph, parent, n1))
+                )
+
+        # take the intersection of both sets to get the minimal confounder set
+        # parents of confounders may be identified as confounders if they have a directed path to the second node
+        # only occurs in one configuration, i.e. either forward or reverse direction
+        minimal_confounders = confounders.intersection(confounders_reverse)
+
+        return list(minimal_confounders)
+
     if not graph.is_dag():
         raise TypeError(f'Expected a DAG, but got a mixed causal graph.')
 
-    # create a copy of the provided graph and prune the children of node_1 and node_2
+    # Confirm node_1 and node_2 are in the graph.
+    if not graph.node_exists(node_1):
+        raise CausalGraphErrors.NodeDoesNotExistError(f'Node not found: {node_1}')
+    if not graph.node_exists(node_2):
+        raise CausalGraphErrors.NodeDoesNotExistError(f'Node not found: {node_2}')
+
+    # Coerce NodeLike to identifier. We already know they are NodeLike as node_exists does this.
+    node_1_id = Node.identifier_from(node_1)
+    node_2_id = Node.identifier_from(node_2)
+
+    # Ensure node_1 != node_2
+    if node_1_id == node_2_id or node_1 == node_2:
+        raise ValueError('node_1 and node_2 cannot be equal. Please provide different nodes / node identifiers.')
+
```

Comment:
I still believe these checks can be abstracted for future use, but I will just do that in an upcoming PR where I add `identify_instruments` and `identify_mediators`

Reply:
Will be done on later PR that would reuse this code.